unified-runtime/source/common/cuda-hip/stream_queue.hpp

```diff
@@ -0,0 +1,344 @@
+  native_type HostSubmitTimeStream{0};
+  // delay_compute_ keeps track of which streams have been recently reused and
+  // their next use should be delayed. If a stream has been recently reused it
+  // will be skipped the next time it would be selected round-robin style. When
+  // skipped, its delay flag is cleared.
+  std::vector<bool> DelayCompute;
+  // keep track of which streams have applied barrier
+  std::vector<bool> ComputeAppliedBarrier;
+  std::vector<bool> TransferAppliedBarrier;
+  ur_context_handle_t_ *Context;
+  ur_device_handle_t_ *Device;
+  std::atomic_uint32_t RefCount;
+  std::atomic_uint32_t EventCount;
+  std::atomic_uint32_t ComputeStreamIndex;
+  std::atomic_uint32_t TransferStreamIndex;
+  unsigned int NumComputeStreams;
+  unsigned int NumTransferStreams;
+  unsigned int LastSyncComputeStreams;
+  unsigned int LastSyncTransferStreams;
+  unsigned int Flags;
+  ur_queue_flags_t URFlags;
+  int Priority;
+  // When ComputeStreamSyncMutex and ComputeStreamMutex both need to be
+  // locked at the same time, ComputeStreamSyncMutex should be locked first
+  // to avoid deadlocks
+  std::mutex ComputeStreamSyncMutex;
+  std::mutex ComputeStreamMutex;
+  std::mutex TransferStreamMutex;
+  std::mutex BarrierMutex;
+  bool HasOwnership;
+
+  stream_queue_t(std::vector<native_type> &&ComputeStreams,
+                 std::vector<native_type> &&TransferStreams,
+                 ur_context_handle_t_ *Context, ur_device_handle_t_ *Device,
+                 unsigned int Flags, ur_queue_flags_t URFlags, int Priority,
+                 bool BackendOwns = true)
+      : ComputeStreams{std::move(ComputeStreams)},
+        TransferStreams{std::move(TransferStreams)},
+        DelayCompute(this->ComputeStreams.size(), false),
+        ComputeAppliedBarrier(this->ComputeStreams.size()),
+        TransferAppliedBarrier(this->TransferStreams.size()), Context{Context},
+        Device{Device}, RefCount{1}, EventCount{0}, ComputeStreamIndex{0},
+        TransferStreamIndex{0}, NumComputeStreams{0}, NumTransferStreams{0},
+        LastSyncComputeStreams{0}, LastSyncTransferStreams{0}, Flags(Flags),
+        URFlags(URFlags), Priority(Priority), HasOwnership{BackendOwns} {
+    urContextRetain(Context);
+  }
+
+  ~stream_queue_t() { urContextRelease(Context); }
+
+  virtual void computeStreamWaitForBarrierIfNeeded(native_type Strean,
+                                                   uint32_t StreamI) = 0;
+  virtual void transferStreamWaitForBarrierIfNeeded(native_type Stream,
+                                                    uint32_t StreamI) = 0;
+  virtual void createStreamWithPriority(native_type *Stream, unsigned int Flags,
+                                        int Priority) = 0;
+  virtual ur_queue_handle_t getEventQueue(const ur_event_handle_t) = 0;
+  virtual uint32_t getEventComputeStreamToken(const ur_event_handle_t) = 0;
+  virtual native_type getEventStream(const ur_event_handle_t) = 0;
+
```

Comment:
I'm wondering if we should be concerning ourselves with the performance overhead that those virtual functions give us and if it's worth considering a static polymorphism design here?

Reply:
That's a very theoretical concern, there's plenty of expensive operations happening in this code, I highly doubt using a few virtual functions will have a noticeable impact on the performance.

Regardless I do intend to improve this code further, I just want to get the restructuring through before touching the functional parts so people can give proper reviews without having the diff mixed up between functional and non-functional change. At that point I'll take another look at these and see if there's a better way to handle them.